miner/collator.go

```diff
@@ -0,0 +1,245 @@
+			continue
+		}
+
+		// Start executing the transaction
+		bs.state.Prepare(tx.Hash(), w.current.tcount)
+		var txLogs []*types.Log
+		txLogs, err = w.commitTransaction(tx, bs.Coinbase())
+		if err == nil {
+			logs = append(logs, txLogs...)
+			tcount++
+		} else {
+			log.Trace("Tx block inclusion failed", "sender", from, "nonce", tx.Nonce(),
+				"type", tx.Type(), "hash", tx.Hash(), "err", err)
+			break
+		}
+	}
+	if err != nil {
+		bs.state.RevertToSnapshot(snap)
+	} else {
+		bs.logs = append(bs.logs, logs...)
+		w.current.tcount = tcount
+	}
+	return err
+}
+
+func (bs *blockState) Gas() (remaining uint64) {
+	return bs.worker.current.gasPool.Gas()
+}
+
+// The DefaultCollator is the 'normal' block collator. It assembles a block
+// based on transaction price ordering.
+type DefaultCollator struct {
+	pool Pool
+}
+
+var (
+	ErrResubmitIntervalElapsed = errors.New("recommit interval elapsed")
+	ErrNewHead                 = errors.New("new chain head received")
+	ErrNoCurrentEnv            = errors.New("missing env for mining")
+)
+
+func (w *DefaultCollator) submit(bs BlockState, txs *types.TransactionsByPriceAndNonce, interrupt *int32) error {
+	for {
+		if interrupt != nil && atomic.LoadInt32(interrupt) != commitInterruptNone {
+			if atomic.LoadInt32(interrupt) == commitInterruptResubmit {
+				return ErrResubmitIntervalElapsed
+			} else {
+				return ErrNewHead
+			}
+		}
+
+		// If we don't have enough gas for any further transactions then we're done
+		available := bs.Gas()
+		if available < params.TxGas {
+			break
+		}
+
+		// Retrieve the next transaction and abort if all done
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+
+		// Enough space for this tx?
+		if available < tx.Gas() {
+			txs.Pop()
+			continue
+		}
+
+		// Error may be ignored here. The error has already been checked
+		// during transaction acceptance is the transaction pool.
+		err := bs.AddTransactions(types.Transactions{tx})
+		switch {
+		case errors.Is(err, core.ErrGasLimitReached):
+			fallthrough
+		case errors.Is(err, core.ErrTxTypeNotSupported):
+			fallthrough
+		case errors.Is(err, core.ErrNonceTooHigh):
+			txs.Pop()
+		case errors.Is(err, core.ErrNonceTooLow):
+			fallthrough
+		case errors.Is(err, nil):
+			fallthrough
+		default:
+			txs.Shift()
+		}
+	}
+
+	return nil
+}
+
+// CollateBlock fills a block based on the highest paying transactions from the
+// transaction pool, giving precedence over 'local' transactions.
+func (w *DefaultCollator) CollateBlock(bs BlockState, txs map[common.Address]types.Transactions, interrupt *int32, isSealing bool) error {
+	if isSealing {
+		// Split the pending transactions into locals and remotes
+		localTxs, remoteTxs := make(map[common.Address]types.Transactions), txs
+		for _, account := range w.pool.Locals() {
+			if accountTxs := remoteTxs[account]; len(accountTxs) > 0 {
+				delete(remoteTxs, account)
+				localTxs[account] = accountTxs
+			}
+		}
+
```

Comment:
In general, I've learned, we don't put an empty line after a cosing bracket. The single bracket is a pretty good "demarcator" in itself, so no extra empty line is needed.